doc/design/api.md

```diff
@@ -0,0 +1,277 @@
+# PaddlePaddle Design Doc
+
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
+- test(cost, ...)
+
+GradientMachine = {Evaluator*, gradients}
+- backward
```

Comment:
Maybe gradient_machine's API could be

```python
result = gradient_machine.forward_backward(..., on_parameter_backwarded=None)  # basically same arguments in `Evaluator.forward`
```

The details show below.

* on_parameter_backwarded will be invoked when the parameter's gradient is calculated immediately. This callback is very useful for network communication. We could send some parameter's gradient while other gradients are calculating.
* Combine forward and backward into one method, because it is not necessarily two stage, especially for multi-threading calculation.
   * The multi-thread backward could not wait for all thread forwarding complete. Each thread could do forward-backward individually. So combine `forward-backward` method together, could let we have a chance to optimize multi-thread training.
   * Not every forwarding can be backward. We must record all activations for backward. But for model inference, forward only need get the final output, intermediate activations could be dropped for saving memory.